packages/builtin/_arithmetic.pony

```diff
@@ -0,0 +1,185 @@
+
+  fun rem_checked[T: UnsignedInteger[T] val](x: T, y: T): (T, Bool) =>
+    (x % y, y == T.from[U8](0))
+
+  fun fld_checked[T: UnsignedInteger[T] val](x: T, y: T): (T, Bool) =>
+    div_checked[T](x, y)
+
+  fun mod_checked[T: UnsignedInteger[T] val](x: T, y: T): (T, Bool) =>
+    rem_checked[T](x, y)
+
+
+primitive _SignedCheckedArithmetic
+
+  fun _mul_checked[U: UnsignedInteger[U] val, T: (Signed & SignedInteger[T, U] val)](x: T, y: T): (T, Bool) =>
+    """
+    basically exactly what the runtime functions __muloti4, mulodi4 etc. are doing
+    and roughly as fast as these.
+
+    Additionally on (at least some) 32 bit systems, the runtime function for checked 64 bit integer addition __mulodi4 is not available.
+    So we shouldn't use: `@"llvm.smul.with.overflow.i64"[(I64, Bool)](this, y)`
+
+    Also see https://bugs.llvm.org/show_bug.cgi?id=14469
+
+    That's basically why we rolled our own.
+    """
+    let result = x * y
+    if x == T.min_value() then
+      return (result, (y != T.from[I8](0)) and (y != T.from[I8](1)))
+    end
+    if y == T.min_value() then
+      return (result, (x != T.from[I8](0)) and (x != T.from[I8](1)))
+    end
+    let x_neg = x >> (x.bitwidth() - U.from[U8](1))
+    let x_abs = (x xor x_neg) - x_neg
+    let y_neg = y >> (x.bitwidth() - U.from[U8](1))
+    let y_abs = (y xor y_neg) - y_neg
+
+    if ((x_abs < T.from[I8](2)) or (y_abs < T.from[I8](2))) then
+      return (result, false)
+    end
+    if (x_neg == y_neg) then
+      (result, (x_abs > (T.max_value() / y_abs)))
+    else
+      (result, (x_abs > (T.min_value() / -y_abs)))
+    end
+
+
+  fun div_checked[T: (SignedInteger[T, U] val & Signed), U: UnsignedInteger[U] val](x: T, y: T): (T, Bool) =>
+    (x / y, (y == T.from[I8](0)) or ((y == T.from[I8](I8(-1))) and (x == T.min_value())))
+
+  fun rem_checked[T: (SignedInteger[T, U] val & Signed), U: UnsignedInteger[U] val](x: T, y: T): (T, Bool) =>
+    (x % y, (y == T.from[I8](0)) or ((y == T.from[I8](I8(-1))) and (x == T.min_value())))
+
+  fun fld_checked[T: (SignedInteger[T, U] val & Signed), U: UnsignedInteger[U] val](x: T, y: T): (T, Bool) =>
+    (x.fld(y), (y == T.from[I8](0)) or ((y == T.from[I8](I8(-1))) and (x == T.min_value())))
+
+  fun mod_checked[T: (SignedInteger[T, U] val & Signed), U: UnsignedInteger[U] val](x: T, y: T): (T, Bool) =>
+    (x.mod(y), (y == T.from[I8](0)) or ((y == T.from[I8](I8(-1))) and (x == T.min_value())))
+
+
+trait _PartialArithmetic
+  fun add_partial[T: (Integer[T] val & Int)](x: T, y: T): T? =>
+    (let r: T, let overflow: Bool) = x.addc(y)
+    if overflow then error else r end
+
+  fun sub_partial[T: (Integer[T] val & Int)](x: T, y: T): T? =>
+    (let r: T, let overflow: Bool) = x.subc(y)
+    if overflow then error else r end
+
+  fun mul_partial[T: (Integer[T] val & Int)](x: T, y: T): T? =>
+    (let r: T, let overflow: Bool) = x.mulc(y)
+    if overflow then error else r end
+
+primitive _UnsignedPartialArithmetic is _PartialArithmetic
+
```

Comment:
Small nitpick: several places in this file, including here, there is an extra empty line between the type declaration and the declaration of the first method.